core/src/banking_simulation.rs

```diff
@@ -0,0 +1,616 @@
+                    GRACE_TICKS_FACTOR * MAX_GRACE_SLOTS,
+                );
+                debug!("{next_leader_slot:?}");
+                poh_recorder
+                    .write()
+                    .unwrap()
+                    .reset(bank.clone_without_scheduler(), next_leader_slot);
+                info!("Bank::new_from_parent()!");
+
+                let old_slot = bank.slot();
+                if let Some(event_time) =
+                    self.banking_trace_events.freeze_time_by_slot.get(&old_slot)
+                {
+                    if log_enabled!(log::Level::Info) {
+                        let current_simulation_time = SystemTime::now();
+                        let elapsed_simulation_time = current_simulation_time
+                            .duration_since(base_simulation_time)
+                            .unwrap();
+                        let elapsed_event_time =
+                            event_time.duration_since(base_event_time).unwrap();
+                        info!(
+                            "jitter(parent_slot: {}): {}{:?} (sim: {:?} event: {:?})",
+                            old_slot,
+                            if elapsed_simulation_time > elapsed_event_time {
+                                "+"
+                            } else {
+                                "-"
+                            },
+                            if elapsed_simulation_time > elapsed_event_time {
+                                elapsed_simulation_time - elapsed_event_time
+                            } else {
+                                elapsed_event_time - elapsed_simulation_time
+                            },
+                            elapsed_simulation_time,
+                            elapsed_event_time,
+                        );
+                    }
+                }
+                bank.freeze();
+                let new_slot = if bank.slot() == parent_slot {
+                    info!("initial leader block!");
+                    self.first_simulated_slot
+                } else {
+                    info!("next leader block!");
+                    bank.slot() + 1
+                };
+                info!("new leader bank slot: {new_slot}");
+                let new_leader = leader_schedule_cache
+                    .slot_leader_at(new_slot, None)
+                    .unwrap();
+                if simulated_leader != new_leader {
+                    info!(
+                        "bank cost: slot: {} {:?} (frozen)",
+                        bank.slot(),
+                        bank.read_cost_tracker()
+                            .map(|t| (t.block_cost(), t.vote_cost()))
+                            .unwrap()
+                    );
+                    info!(
+                        "{} isn't leader anymore at slot {}; new leader: {}",
+                        simulated_leader, new_slot, new_leader
+                    );
+                    break;
```

Comment:
So when we're no longer leader the process will end.

In the future, could we extend this capability so that we "fast-forward" through our non-leader periods?

That probably adds considerable complexity, but I think would make simming significantly more useful.
AFAICT, as is we must load from snapshot everytime we want to do a sim of 4 slots - which will be very time consuming if I have hundred(s) of leader slots in my trace data that I'd like to simulate.

Reply:
Think there's some complexity in how we'd need to "fast-forward" through non-leader periods, but don't think it'd require loading from a snapshot if done correctly.

Probably would add even more complexity, but could we not treat the simmed blocks as some sort of duplicate block (or a fork).
After each 4 slot sim, we drop the simmed blocks (duplicates/fork) for the **actual** blocks which we then replay as normal until we get close to next leader period.

That's what I had in mind for the fast-forward, since I definitely agree that we can't just continue on from the simmed blocks and act like things will just work from there haha.
If we were to do this would probably want to collaborate with ashwin or stevecz on how we could handle the "sim-swap" to remove simmed blocks and insert real blocks.

Reply:
oh, that idea sounds nice.

Reply:
> how we could handle the "sim-swap" to remove simmed blocks and insert real blocks.

i think this just can be done with read-write side blockstore.

Reply:
> oh, that idea sounds nice.

Cool! I think that'd really improve the usability, but we should definitely leave it for a follow-up PR, this one is big enough as is 😄